packages/snap/src/encryption.ts

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: BUSL-1.1
+
+import { ENCRYPTION_VERSION } from '@galactica-net/galactica-types';
+import {
+  EncryptedZkCert,
+  ImportZkCertError,
+  ZkCertRegistered,
+} from '@galactica-net/snap-api';
+import {
+  getEncryptionPublicKey,
+  encryptSafely,
+  decryptSafely,
+} from '@metamask/eth-sig-util';
+import { SnapsGlobalObject } from '@metamask/rpc-methods';
+import { Buffer } from 'buffer';
+
+import { checkZkCert } from './zkCertHandler';
+
+/**
+ * Create a new encryption key pair for the holder. It is used to encrypt personal details in ZK certificates, for example on the way from guardian to the holder.
+ *
+ * @param snap - The snap for interaction with Metamask.
+ * @returns The public and private key.
+ */
+export async function createEncryptionKeyPair(snap: SnapsGlobalObject) {
+  // It is derived from the user's private key handled by Metamask. Meaning that HW wallets are not supported.
+  // The plan to support HW wallets is to use the `eth_sign` method to derive the key from a signature.
+  // However this plan is currently not supported anymore as discussed here: https://github.com/MetaMask/snaps/discussions/1364#discussioncomment-5719039
+  const entropy = await snap.request({
+    method: 'snap_getEntropy',
+    params: {
+      version: 1,
+      salt: 'galactica-encryption2',
+    },
+  });
+  const privKey = entropy.slice(2); // remove 0x prefix
+  prepareGlobalsForEthSigUtil();
+  const publicKey = getEncryptionPublicKey(privKey);
+  return { pubKey: publicKey, privKey };
+}
+
+/**
+ * Encrypt a zkCert for exporting.
+ *
+ * @param zkCert - The ZkCertRegistered to encrypt.
+ * @param pubKey - The public key for encryption.
+ * @param holderCommitment - The holder commitment to associate the zkCert with the holder who can decrypt it.
+ * @returns The encrypted ZkCertRegistered as EthEncryptedData.
+ */
+export function encryptZkCert(
+  zkCert: ZkCertRegistered,
+  pubKey: string,
+  holderCommitment: string,
+): EncryptedZkCert {
+  const message = JSON.stringify(zkCert);
+  prepareGlobalsForEthSigUtil();
+  const encryptedZkCert = encryptSafely({
+    publicKey: pubKey,
+    data: message,
+    version: ENCRYPTION_VERSION,
+  }) as EncryptedZkCert;
```

Comment:
Hi, Alex C. found out that, encryptSafely already performs JSON conversion, so we can omit line 55 if that is the case.

Reply:
Changed it